### PR TITLE
Fixing tests on GPU

### PR DESCRIPTION
**Pull request by jscanvic**
Towards #556 

**Testing command**

```bash
DEEPINV_MOCK_TESTS=True python -m pytest -v deepinv/tests 2>&1 > >(tee test.log)
```

**Results before fixes**

[full logs](https://github.com/user-attachments/files/20909794/test.log)

```
=484 failed, 701 passed, 99 skipped, 137 warnings, 363 errors in 559.19s (0:09:19) =
```

### Checks to be done before submitting your PR
- [ ] `python3 -m pytest tests/` runs successfully.
- [ ] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

**Comment by jscanvic** [comment]
We're almost there but there's a last bug I can't manage to fix - I'll leave it for another PR.

File: deepinv/loss/metric/distortion.py [hunk 1]
```diff
@@ -680,16 +680,16 @@ def metric(self, x_net: Tensor = None, x: Tensor = None, *args, **kwargs) -> Ten
         coeffs_deg_y = self._haar_wavelet_decompose(deg_y, n_scales)
         if is_color_image:
             coefficients_ref_i = torch.abs(
-                self._convolve2d(ref_i, torch.ones((2, 2)) / 4.0)
+                self._convolve2d(ref_i, torch.ones((2, 2), device=ref_i.device) / 4.0)
```
Comment (mh-nguyen712) [inline deepinv/loss/metric/distortion.py]:
```suggestion
                self._convolve2d(ref_i, torch.ones((2, 2), device=ref_i.device, dtype=ref_i.dtype) / 4.0)
```

Comment (mh-nguyen712) [reply deepinv/loss/metric/distortion.py]: and other lines as well please

File: deepinv/physics/generator/base.py [hunk 2]
```diff
@@ -115,7 +124,8 @@ def reset_rng(self):
         r"""
         Reset the random number generator to its initial state.
         """
-        self.rng.set_state(self.initial_random_state)
+        # NOTE: Generator.set_state expects a tensor living on the CPU.
+        self.rng.set_state(self.initial_random_state.cpu())
```
Comment (mh-nguyen712) [inline deepinv/physics/generator/base.py:128]: Thanks!

File: deepinv/physics/noise.py [hunk 3]
```diff
@@ -22,6 +22,14 @@ def __init__(self, noise_model: Callable = None, rng: torch.Generator = None):
         self.noise_model = noise_model
         self.rng = rng
         if rng is not None:
+            # NOTE: There is no use in moving RNG states from one device to another
```
Comment (mh-nguyen712) [inline deepinv/physics/noise.py:25]: Another option could be:
- save the `initial_random_noise` as a regular attribute `self.initial_random_noise = rng.get_state()`
- then over-write the `self.state_dict()` and `self.load_state_dict()` method.
What do you think?

File: deepinv/tests/test_models.py [hunk 4]
```diff
@@ -493,14 +493,22 @@ def test_wavelet_decomposition(channels, dimension, batch_size, device):
     # 1 decomposition
     out = model.dwt(x)
     x_hat = model.iwt(out)
-    assert x_hat.shape == x.shape and torch.allclose(x, x_hat, rtol=1e-5, atol=1e-5)
+
+    # For some reason the precision is more than 100x lower on GPU.
+    tol = 1e-3 if torch.device(device).type == "cuda" else 1e-5
```
Comment (mh-nguyen712) [inline deepinv/tests/test_models.py:498]: It passed on my GPU so it might by hardware dependent?

**Review by mh-nguyen712 — COMMENTED** [review]
Many thanks Jérémy. Just have some minor comments

File: deepinv/physics/noise.py (hunk 3 above)
Comment (jscanvic) [reply deepinv/physics/noise.py:25]: I like the idea but I'm worried of breaking something!

Do you think this would work?

```python
def state_dict(self, *args, **kwargs):
    dict = super().state_dict(*args, **kwargs)
    dict["initial_random_noise"] = self.initial_random_noise
    return dict

def load_state_dict(self, state_dict, *args, **kwargs):
    self.initial_random_noise = dict.popitem("initial_random_noise")
    super().load_state_dict(state_dict)

```

File: deepinv/tests/test_models.py (hunk 4 above)
Comment (jscanvic) [reply deepinv/tests/test_models.py:498]: Oh okay, let me revert it then!

Comment (jscanvic) [reply deepinv/tests/test_models.py:498]: The last error I had was related to this specific test - I think it's not just a tolerance issue anyway

File: deepinv/physics/noise.py (hunk 3 above)
Comment (mh-nguyen712) [reply deepinv/physics/noise.py:25]: It could work but I just realized that we also have to put `self.rng.set_state(self.initial_rng_state.cpu())`, to prevent error if someone load the state_dict with `torch. Load(..., map_device='cuda')`

Comment (mh-nguyen712) [reply deepinv/physics/noise.py:25]: So I think it does not help much...  So a simple call `.cpu()` should be ok since this functions is not called regularly...

File: deepinv/tests/test_models.py (hunk 4 above)
Comment (mh-nguyen712) [reply deepinv/tests/test_models.py:498]: Just checked it again on my side. It passed on a RTX5000 but not on my laptop GPU with tol=1e-5

File: deepinv/physics/noise.py (hunk 3 above)
Comment (jscanvic) [reply deepinv/physics/noise.py:25]: Hmm right! Let's keep it that way then.

File: deepinv/tests/test_models.py (hunk 4 above)
Comment (jscanvic) [reply deepinv/tests/test_models.py:498]: Weird... What do you suggest we do?

Comment (mh-nguyen712) [reply deepinv/tests/test_models.py:498]: I think we could set tol=1e-3 for GPU as you did before you contributors dont have issues when testing. Sorry for the inconvenience.

Comment (jscanvic) [reply deepinv/tests/test_models.py:498]: I made the change! Should be good now.
